src/example_09.rs

```diff
@@ -47,17 +47,22 @@ mod game {
         }
         pub fn run(&mut self) {
             while self.reserve.value() != 0 {
-                let mut player = self.player_queue.pop_front().unwrap();
-                // in theory here it would be convenient to have all players
-                // share a mutable reference to the reserve so that it is not
-                // necessary to pass it around everytime
-                // however I think that rust does not allows it because of the
-                // ownership rules. Wondering if there is a smart way to
-                // have both benefits
-                player.play(&mut self.reserve);
-                self.player_queue.push_back(player);
```

Comment:
Okay so where we have the Game struct that keep a list of players and a reserve.

during the game I want to modify at the same time one of the players and the reserve.

in the previous commit I was taking ownership of the player, and borrowing only the reserve  so that I could change both.

Now I'm trying to refactor it because I felt like taking something out and putting it back feels like breaking the encapsulation principle.

But It looks like if I borrow two parts of Game at the same time it is like if I'm borrowing mutable references of Game twice.


Is there a smarter way to keep things tidy and sane at the same time?
I understand the principle of borrowing a mutable instance only once but in practice this makes your life miserable :sweat: 
Am I missing something?

Reply:
here is the error I get if I try to compile

``` 
➜  rust_examples git:(main) ✗ cargo fmt && cargo run
   Compiling rust_examples v0.1.0 (/home/corsair/code/rust/rust_examples)
error[E0499]: cannot borrow `*self` as mutable more than once at a time
  --> src/example_09.rs:51:31
   |
50 |                 let player = self.next_player();
   |                              ------------------ first mutable borrow occurs here
51 |                 let reserve = self.mut_reserve();
   |                               ^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
52 |
53 |                 player.play(reserve);
   |                 -------------------- first borrow later used here

For more information about this error, try `rustc --explain E0499`.
error: could not compile `rust_examples` (bin "rust_examples") due to previous error
```